lib/libvarnishapi/vsl_dispatch.c

```diff
@@ -1299,22 +1338,26 @@ vslq_next(struct VSLQ *vslq)
 		    c->rec.ptr;
 		if (len == 0)
 			return (r);
-		tag = (enum VSL_tag_e)VSL_TAG(VSL_NEXT(c->rec.ptr));
+		ptr = VSL_NEXT(c->rec.ptr);
+		tag = (enum VSL_tag_e)VSL_TAG(ptr);
 	} else {
 		vxid = VSL_ID(c->rec.ptr);
 		len = VSL_NEXT(c->rec.ptr) - c->rec.ptr;
+		ptr = c->rec.ptr;
 	}
 	assert(len > 0);
 	if (vxid == 0)
 		/* Skip non-transactional records */
 		return (r);
 
 	vtx = vtx_lookup(vslq, vxid);
-	if (vtx == NULL && tag == SLT_Begin) {
+	keep = tag != SLT_Begin || vslq_candidate(vslq, ptr);
```

Comment:
Should the test of tag shouldn't be moved into vslq_candidate as well, to keep all the logic in one place?